File: tools/rag-dotnet/src/RagTools.Core/RemoteIngestClient.cs
using System.IO.Compression;
using System.Text.Json;
using System.Text.Json.Serialization;
using Microsoft.Extensions.Logging;

namespace RagTools.Core;

/// <summary>
/// Pushes markdown files to a remote mcp_server ingest endpoint via HTTP.
///
/// Wire protocol:
///   1. Build a single ZIP containing rag-config.yaml + metadata-rules.yaml + queries.yaml
///      at the ZIP root, plus every file-to-process at its workspace-relative path.
///   2. POST /ingest/{collection}/batch with Content-Type application/zip.
///   3. Parse batch_id + operations[] from the 202 response.
///   4. Poll GET /ingest/{collection}/operations/{operation_id} until every
///      operation is Completed or Failed.
///
/// Mirrors the Python remote_ingest_client.push_files_to_remote_server().
/// </summary>
public sealed class RemoteIngestClient : IDisposable
{
    private readonly HttpClient _http;
    private readonly ILogger _log;
    private static readonly JsonSerializerOptions JsonOpts = new(JsonSerializerDefaults.Web);

    public RemoteIngestClient(string baseUrl, string? apiKey, ILogger log)
    {
        _log = log;
        _http = new HttpClient
        {
            BaseAddress = new Uri(baseUrl.TrimEnd('/') + "/"),
            Timeout = TimeSpan.FromMinutes(5),
        };
        if (!string.IsNullOrEmpty(apiKey))
            _http.DefaultRequestHeaders.Add("X-Api-Key", apiKey);
    }

    // Server queue capacity is 100; stay safely below.
    private const int BatchChunkSize = 80;

    public async Task<int> PushAsync(
        RagConfig cfg,
        IReadOnlyList<(FileInfo File, string RelPath, string Hash)> toProcess,
        ManifestService manifest,
        CancellationToken ct = default)
    {
        if (toProcess.Count == 0)
        {
            _log.LogInformation("remote push: no files to send");
            return 0;
        }

        if (toProcess.Count <= BatchChunkSize)
            return await PushOneBatchAsync(cfg, toProcess, manifest, "1/1", ct);

        var chunks = new List<IReadOnlyList<(FileInfo, string, string)>>();
        for (var i = 0; i < toProcess.Count; i += BatchChunkSize)
        {
            var take = Math.Min(BatchChunkSize, toProcess.Count - i);
            var slice = new (FileInfo, string, string)[take];
            for (var j = 0; j < take; j++) slice[j] = toProcess[i + j];
            chunks.Add(slice);
        }
        _log.LogInformation("splitting {Total} file(s) into {N} batch(es) of ≤{Size}",
            toProcess.Count, chunks.Count, BatchChunkSize);

        var totalFailed = 0;
        for (var i = 0; i < chunks.Count; i++)
        {
            var failed = await PushOneBatchAsync(
                cfg, chunks[i], manifest, $"{i + 1}/{chunks.Count}", ct);
            totalFailed += failed;
        }
        _log.LogInformation("remote ingest: {Ok} ok, {Failed} failed ({Total} total)",
            toProcess.Count - totalFailed, totalFailed, toProcess.Count);
        return totalFailed;
    }

    private async Task<int> PushOneBatchAsync(
        RagConfig cfg,
        IReadOnlyList<(FileInfo File, string RelPath, string Hash)> toProcess,
        ManifestService manifest,
        string label,
        CancellationToken ct)
    {
        var collection = cfg.Collection;
        byte[] zipBytes;
        try
        {
            zipBytes = BuildBatchZip(cfg, toProcess);
        }
        catch (Exception ex)
        {
            _log.LogError(ex, "failed to build batch ZIP");
            return toProcess.Count;
        }

        _log.LogInformation(
            "batch {Label} ZIP ready ({Bytes} bytes), POST /ingest/{Collection}/batch ({Count} file(s))",
            label, zipBytes.Length, collection, toProcess.Count);

        BatchAcceptedResponse? batch = null;
        const int MaxAttempts = 8; // ~2 min of 503 retries (server queue drain)
        for (var attempt = 0; attempt < MaxAttempts; attempt++)
        {
            using var content = new ByteArrayContent(zipBytes);
            content.Headers.ContentType = new System.Net.Http.Headers.MediaTypeHeaderValue("application/zip");
            try
            {
                using var resp = await _http.PostAsync(
                    $"ingest/{Uri.EscapeDataString(collection)}/batch", content, ct);

                if (resp.StatusCode == System.Net.HttpStatusCode.ServiceUnavailable && attempt < MaxAttempts - 1)
                {
                    var wait = TimeSpan.FromSeconds(Math.Min(30, Math.Pow(2, attempt)));
                    _log.LogWarning("batch POST 503 queue full — retry in {Wait}s", wait.TotalSeconds);
                    await Task.Delay(wait, ct);
                    continue;
                }

                if (!resp.IsSuccessStatusCode)
                {
                    var errBody = await resp.Content.ReadAsStringAsync(ct);
                    _log.LogError("batch POST failed HTTP {Code} — {Body}",
                        (int)resp.StatusCode, errBody.Length > 400 ? errBody[..400] : errBody);
                    return toProcess.Count;
                }

                var json = await resp.Content.ReadAsStringAsync(ct);
                batch = JsonSerializer.Deserialize<BatchAcceptedResponse>(json, JsonOpts);
                break;
            }
            catch (Exception ex)
            {
                _log.LogError(ex, "batch POST exception (attempt {N})", attempt);
                if (attempt >= MaxAttempts - 1) return toProcess.Count;
                await Task.Delay(TimeSpan.FromSeconds(2), ct);
            }
        }

        if (batch?.Operations is null || batch.Operations.Count == 0)
        {
            _log.LogError("server returned no operations");
            return toProcess.Count;
        }

        var relByOp = batch.Operations
            .Where(o => !string.IsNullOrEmpty(o.OperationId))
            .ToDictionary(o => o.OperationId!, o => o.RelPath ?? "?");
        _log.LogInformation("batch accepted — {Count} operation(s), polling …", relByOp.Count);

        var hashByRel = toProcess.ToDictionary(t => t.RelPath, t => t.Hash);

        var remaining = new HashSet<string>(relByOp.Keys);
        var failed = 0;
        var deadline = DateTime.UtcNow + TimeSpan.FromSeconds(Math.Max(180, remaining.Count * 5));

        while (remaining.Count > 0 && DateTime.UtcNow < deadline)
        {
            await Task.Delay(TimeSpan.FromSeconds(2), ct);
            foreach (var opId in remaining.ToList())
            {
                OperationStatusResponse? op = null;
                try
                {
                    using var resp = await _http.GetAsync(
                        $"ingest/{Uri.EscapeDataString(collection)}/operations/{Uri.EscapeDataString(opId)}", ct);
                    if (resp.StatusCode == System.Net.HttpStatusCode.NotFound)
                    {
                        _log.LogError("FAIL {Rel}: operation 404", relByOp[opId]);
                        failed++;
                        remaining.Remove(opId);
                        continue;
                    }
                    if (!resp.IsSuccessStatusCode) continue;
                    var json = await resp.Content.ReadAsStringAsync(ct);
                    op = JsonSerializer.Deserialize<OperationStatusResponse>(json, JsonOpts);
                }
                catch { continue; }

                var status = op?.Status?.Trim() ?? "";
                if (status.Equals("Completed", StringComparison.OrdinalIgnoreCase) ||
                    status.Equals("Complete", StringComparison.OrdinalIgnoreCase))
                {
                    var rel = relByOp[opId];
                    if (hashByRel.TryGetValue(rel, out var hash))
                        manifest.Update(rel, hash, op?.ChunkCount ?? 0);
                    _log.LogInformation("OK  {Rel} → {Chunks} chunk(s)", rel, op?.ChunkCount ?? 0);
                    remaining.Remove(opId);
                }
                else if (status.Equals("Failed", StringComparison.OrdinalIgnoreCase))
                {
                    _log.LogError("FAIL {Rel}: {Err}", relByOp[opId],
                        op?.ErrorMessage ?? "unknown error");
                    failed++;
                    remaining.Remove(opId);
                }
            }
        }

        foreach (var opId in remaining)
        {
            _log.LogError("TIMEOUT {Rel}: operation did not complete in time", relByOp[opId]);
            failed++;
        }

        var ok = relByOp.Count - failed;
        _log.LogInformation("remote ingest: {Ok} ok, {Failed} failed ({Total} total)",
            ok, failed, relByOp.Count);
        return failed;
    }

    private static byte[] BuildBatchZip(
        RagConfig cfg,
        IReadOnlyList<(FileInfo File, string RelPath, string Hash)> files)
    {
        if (string.IsNullOrEmpty(cfg.LoadedFrom))
            throw new InvalidOperationException("RagConfig.LoadedFrom is required for remote batch upload");
        if (string.IsNullOrEmpty(cfg.MetadataRulesPath) || !File.Exists(cfg.MetadataRulesPath))
            throw new InvalidOperationException(
                "metadata-rules.yaml not found — required for batch upload");
        if (string.IsNullOrEmpty(cfg.QueriesPath) || !File.Exists(cfg.QueriesPath))
            throw new InvalidOperationException(
                "queries.yaml not found — required for batch upload");

        using var ms = new MemoryStream();
        using (var zip = new ZipArchive(ms, ZipArchiveMode.Create, leaveOpen: true))
        {
            AddEntry(zip, "rag-config.yaml", File.ReadAllBytes(cfg.LoadedFrom));
            AddEntry(zip, "metadata-rules.yaml", File.ReadAllBytes(cfg.MetadataRulesPath));
            AddEntry(zip, "queries.yaml", File.ReadAllBytes(cfg.QueriesPath));
            if (!string.IsNullOrEmpty(cfg.GlossaryPath) && File.Exists(cfg.GlossaryPath))
                AddEntry(zip, "multilingual-glossary.yaml", File.ReadAllBytes(cfg.GlossaryPath));
            foreach (var (fi, rel, _) in files)
            {
                try { AddEntry(zip, rel.Replace('\\', '/'), File.ReadAllBytes(fi.FullName)); }
                catch (IOException) { /* skip unreadable */ }
            }
        }
        return ms.ToArray();
    }

    private static void AddEntry(ZipArchive zip, string entryName, byte[] data)
    {
        var entry = zip.CreateEntry(entryName, CompressionLevel.Optimal);
        using var es = entry.Open();
        es.Write(data, 0, data.Length);
    }

    public void Dispose() => _http.Dispose();

    private sealed class BatchAcceptedResponse
    {
        [JsonPropertyName("batch_id")] public string? BatchId { get; set; }
        [JsonPropertyName("count")] public int Count { get; set; }
        [JsonPropertyName("operations")] public List<BatchOperationStub>? Operations { get; set; }
    }

    private sealed class BatchOperationStub
    {
        [JsonPropertyName("operation_id")] public string? OperationId { get; set; }
        [JsonPropertyName("rel_path")] public string? RelPath { get; set; }
    }

    private sealed class OperationStatusResponse
    {
        [JsonPropertyName("status")] public string? Status { get; set; }
        [JsonPropertyName("chunk_count")] public int? ChunkCount { get; set; }
        [JsonPropertyName("error_message")] public string? ErrorMessage { get; set; }
    }
}

File: tools/rag/remote_ingest_client.py
"""HTTP client for the remote ingest API exposed by mcp_server.py (SSE / HTTP mode).

Mirrors the .NET ``RagTools.Ingest --remote`` flag pattern.

Design notes:
- Pure stdlib: urllib.request / urllib.error + zipfile — no extra dependency.
- Builds a single ZIP containing rag-config.yaml + metadata-rules.yaml + queries.yaml
  at the ZIP root, plus every file to ingest at its repo-relative path.
- POSTs the ZIP once to  POST /ingest/{collection}/batch  (Content-Type application/zip).
- Polls GET /ingest/{collection}/operations/{opId} until every op is Completed or Failed.
- Retries the batch POST on HTTP 503 (queue full) up to 3 times with exponential back-off.
- Returns 0 on full success, 1 if any file failed.

History: prior version POSTed each file to  POST /ingest/{collection}  (per-file JSON).
That route was removed when the batch ZIP endpoint became canonical, so the per-file
version returned HTTP 406 for every file on the current servers.
"""
from __future__ import annotations

import io
import json
import os
import time
import urllib.error
import urllib.request
import zipfile
from pathlib import Path
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from common import Config


def _resolve_companion_paths(cfg: "Config") -> tuple["Path | None", "Path | None"]:
    """Re-resolve metadata-rules.yaml and queries.yaml paths the same way load_config did."""
    from common import _find_companion_file  # type: ignore[attr-defined]

    if env_meta := os.environ.get("RAG_METADATA"):
        meta_path: "Path | None" = Path(env_meta)
    elif cfg.config_path is not None:
        meta_path = _find_companion_file(
            cfg.config_path, "metadata_rules", "metadata-rules.yaml", raw=cfg.raw)
    else:
        meta_path = None

    if env_q := os.environ.get("RAG_QUERIES"):
        q_path: "Path | None" = Path(env_q)
    elif cfg.config_path is not None:
        q_path = _find_companion_file(
            cfg.config_path, "queries", "queries.yaml", raw=cfg.raw)
    else:
        q_path = None

    return meta_path, q_path


def _build_batch_zip(cfg: "Config", files: "list[Path]") -> bytes:
    """Build the batch ZIP payload expected by /ingest/{coll}/batch.

    Layout (all at ZIP root or under their workspace-relative path):
        rag-config.yaml
        metadata-rules.yaml        (referenced by config_files.metadata_rules)
        queries.yaml               (referenced by config_files.queries)
        <rel_path>                 for every file in `files`
    """
    if cfg.config_path is None:
        raise RuntimeError("cfg.config_path is required for remote batch upload")

    meta_path, q_path = _resolve_companion_paths(cfg)
    if meta_path is None or not meta_path.exists():
        raise RuntimeError(
            "metadata-rules.yaml not found — required for batch upload "
            "(set RAG_METADATA or add config_files.metadata_rules to rag-config.yaml)"
        )
    if q_path is None or not q_path.exists():
        raise RuntimeError(
            "queries.yaml not found — required for batch upload "
            "(set RAG_QUERIES or add config_files.queries to rag-config.yaml)"
        )

    buf = io.BytesIO()
    with zipfile.ZipFile(buf, "w", zipfile.ZIP_DEFLATED) as z:
        z.writestr("rag-config.yaml", cfg.config_path.read_bytes())
        z.writestr("metadata-rules.yaml", meta_path.read_bytes())
        z.writestr("queries.yaml", q_path.read_bytes())
        gloss = getattr(cfg, "glossary_path", None)
        if gloss is not None and Path(gloss).exists():
            z.writestr("multilingual-glossary.yaml", Path(gloss).read_bytes())
        for path in files:
            try:
                rel = path.relative_to(cfg.workspace).as_posix()
            except ValueError:
                rel = path.name
            try:
                z.writestr(rel, path.read_bytes())
            except OSError as exc:
                print(f"[ingest] SKIP {rel}: cannot read — {exc}")
    return buf.getvalue()


# Server queue capacity is 100 (ingest_worker.DEFAULT_CAPACITY); stay safely below.
BATCH_CHUNK_SIZE = 80


def push_files_to_remote_server(
    cfg: "Config",
    files: "list[Path]",
    base_url: str,
    api_key: "str | None" = None,
) -> int:
    """POST batch ZIP(s) of files to a running mcp_server ingest endpoint.

    Splits ``files`` into chunks of ``BATCH_CHUNK_SIZE`` so each batch fits under
    the server's queue capacity. Returns 0 on full success, 1 on any failure.
    """
    if not files:
        print("[ingest] remote push: no files to send")
        return 0

    total = len(files)
    if total <= BATCH_CHUNK_SIZE:
        return _push_one_batch(cfg, files, base_url, api_key, label=f"1/1 ({total})")

    chunks = [files[i:i + BATCH_CHUNK_SIZE] for i in range(0, total, BATCH_CHUNK_SIZE)]
    print(f"[ingest] splitting {total} file(s) into {len(chunks)} batch(es) of ≤{BATCH_CHUNK_SIZE}")
    overall = 0
    for idx, chunk in enumerate(chunks, 1):
        rc = _push_one_batch(
            cfg, chunk, base_url, api_key,
            label=f"{idx}/{len(chunks)} ({len(chunk)})",
        )
        if rc != 0:
            overall = rc
    return overall


def _push_one_batch(
    cfg: "Config",
    files: "list[Path]",
    base_url: str,
    api_key: "str | None",
    label: str,
) -> int:
    base = base_url.rstrip("/")
    collection = cfg.collection
    key = (api_key or os.environ.get("RAG_API_KEY", "")).strip()

    print(f"[ingest] building batch {label} ZIP → {collection}")
    try:
        zip_bytes = _build_batch_zip(cfg, files)
    except Exception as exc:  # noqa: BLE001
        print(f"[ingest] ERROR building batch ZIP: {exc}")
        return 1
    print(
        f"[ingest] batch ZIP ready ({len(zip_bytes)} bytes), "
        f"POST → {base}/ingest/{collection}/batch"
    )

    headers: dict[str, str] = {"Content-Type": "application/zip"}
    if key:
        headers["X-Api-Key"] = key

    batch_response: dict = {}
    max_attempts = 8  # ~1+2+4+8+16+30+30+30 = ~2 min of 503 retries (queue drain)
    for attempt in range(max_attempts):
        req = urllib.request.Request(
            f"{base}/ingest/{collection}/batch",
            data=zip_bytes,
            headers=headers,
            method="POST",
        )
        try:
            with urllib.request.urlopen(req, timeout=120) as resp:
                batch_response = json.loads(resp.read())
            break
        except urllib.error.HTTPError as exc:
            body = ""
            try:
                body = exc.read().decode("utf-8", "replace")[:300]
            except Exception:
                pass
            if exc.code == 503 and attempt < max_attempts - 1:
                wait = min(30, 2 ** attempt)
                print(f"[ingest] batch POST: queue full (503), retry in {wait}s … {body}")
                time.sleep(wait)
                continue
            print(f"[ingest] ERROR batch POST failed — HTTP {exc.code}: {body}")
            return 1
        except Exception as exc:  # noqa: BLE001
            print(f"[ingest] ERROR batch POST failed — {exc}")
            return 1
    else:
        return 1

    operations = batch_response.get("operations") or []
    if not operations:
        print(f"[ingest] WARNING server returned no operations: {batch_response!r}")
        return 1

    op_ids = [op["operation_id"] for op in operations if op.get("operation_id")]
    rel_by_id = {op["operation_id"]: op.get("rel_path", "?") for op in operations}
    print(f"[ingest] batch accepted — {len(op_ids)} operation(s), polling …")

    poll_headers: dict[str, str] = {}
    if key:
        poll_headers["X-Api-Key"] = key

    succeeded = failed = 0
    deadline = time.time() + max(180.0, len(op_ids) * 5.0)
    remaining = set(op_ids)
    last_status: dict[str, str] = {}

    while remaining and time.time() < deadline:
        time.sleep(2)
        for op_id in list(remaining):
            poll_req = urllib.request.Request(
                f"{base}/ingest/{collection}/operations/{op_id}",
                headers=poll_headers,
            )
            try:
                with urllib.request.urlopen(poll_req, timeout=15) as resp:
                    op = json.loads(resp.read())
            except urllib.error.HTTPError as exc:
                if exc.code == 404:
                    last_status[op_id] = "NotFound"
                    print(f"[ingest] FAIL {rel_by_id[op_id]}: operation 404")
                    failed += 1
                    remaining.discard(op_id)
                continue
            except Exception:
                continue
            status = (op.get("status") or "").strip()
            last_status[op_id] = status
            if status.lower() in ("completed", "complete"):
                chunks = op.get("chunk_count") or op.get("chunkCount") or "?"
                print(f"[ingest] OK  {rel_by_id[op_id]} → {chunks} chunk(s)")
                succeeded += 1
                remaining.discard(op_id)
            elif status.lower() == "failed":
                err = op.get("error_message") or op.get("errorMessage") or "unknown error"
                print(f"[ingest] FAIL {rel_by_id[op_id]}: {err}")
                failed += 1
                remaining.discard(op_id)

    if remaining:
        for op_id in remaining:
            print(
                f"[ingest] TIMEOUT {rel_by_id[op_id]}: still {last_status.get(op_id, 'Queued')!r}"
            )
            failed += 1

    print(
        f"[ingest] remote push complete: {succeeded} ok, {failed} failed, 0 skipped "
        f"({len(files)} total)"
    )
    return 0 if failed == 0 else 1
